README.md

```diff
@@ -1,2 +1,39 @@
 # rickshaw
 A project to facilitate execution of benchmarks and tools via extensions for many target environments
+
+### Input
+
+Rickshaw needs the following:
+- Benchmark config file
+  - Rickshaw really does not care what the benchmark is; however, rickshaw must have a config file to know how to run the benchmark.  These config files are not bundled with the rickshaw project, but some from other benchmark-helper projects, or you write this config file yourself for your own benchmark.  The config file instructs rickshaw on what to execute on the controller (the host where you are running rickshaw as well as what to run on any benchmark clients and servers. 
+- Benchmark parameters
+  - This may be an array of multiple parameters, when the user wants the benchmark run mulitple ways, for example:
+    <pre>[ "--rw=read --bs=4k", "--rw=write --bs=4k" ]</pre>
+  - The [multiplex](https://github.com/perftool-incubator/multiplex) project can be used to generate this array (it can convert things like "--rw=read,write --bs=4k" into the array above), and it will do parameter validation for you as well.
+- Endpoint extension
+  - This determines how the benchmark gets executed on different endpoints.  The default extension, local, simply runs the benchmark command on the local host.  Rickshaw supports using multiple extensions for different endpoints.  For example, if you want to run uperf benchmark, you need both a client and server uperf.  If you want to run the uperf server on Kuberbetes, but you want to run the uperf client on a baremetal host, you can use the 'k8s' extension for the server and the 'ssh' extension for the client
+  - <pre>--endpoint:server[1]:k8s:$master-hostname --endpoint:client[1]:ssh:$client-hostname</pre>
+  - Other examples
+    - 8 servers running in 8 containers in k8s and 8 clients running on the same barmetal host:  
+      - <pre>--endpoint:server[1-8]:k8s:$master-hostname --endpoint:client[1-8]:ssh:$client-hostname</pre>
```

Comment:
Should this be:

```
--endpoint:server[1-8]:k8s:$cluster-master-hostname --endpoint:client[1-8]:ssh:$client-hostname
```

Reply:
actually I think I should have used: 
<pre> --endpoint:k8s:server[1-8]:$cluster-master-hostname --endpoint:ssh:client[1-8]:$client-hostname </pre>

So the format would be:  --endpoint:<extension>:$type[$range], where $type is client, server or blank, and $range is n-m or if blank, assumed to be the first client/server.

Reply:
As I suggested above, how about:

```
--endpoint:${endpoint-type}:${endpoint-subtype}[$range]:{$type[$range]}

Where ${endpoint-type} could be one of "local", "ssh", "k8s", etc.

Where ${endpoint-subtype} would be context sensitive for the ${endpoint-type}, for "k8s" it could be "pods" otherwise assumed to be all in one pod

The rest as you said.

```